File: app/src/main/java/com/example/lovelink/LoginActivity.java
package com.example.lovelink;

import android.app.Activity;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;

public class LoginActivity extends Activity {

    private EditText emailEditText, passwordEditText;
    private Button loginButton, signupButton;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        // Referencias a los campos de entrada
        emailEditText = findViewById(R.id.emailEditText);
        passwordEditText = findViewById(R.id.passwordEditText);
        loginButton = findViewById(R.id.loginButton);
        signupButton = findViewById(R.id.signupButton);

        // Acción del botón de iniciar sesión
        loginButton.setOnClickListener(v -> {
            String email = emailEditText.getText().toString().trim();
            String password = passwordEditText.getText().toString().trim();

            if (email.isEmpty() || password.isEmpty()) {
                Toast.makeText(LoginActivity.this, "Por favor, ingresa correo y contraseña", Toast.LENGTH_SHORT).show();
                return;
            }

            // Verificar si las credenciales son correctas
            boolean isAuthenticated = authenticateUser(email, password);

            if (isAuthenticated) {
                // Guardar que el usuario ha iniciado sesión
                SharedPreferences.Editor editor = getSharedPreferences("LoveLinkPrefs", MODE_PRIVATE).edit();
                editor.putBoolean("isLoggedIn", true);
                editor.putString("email", email);
                editor.apply();

                // Redirigir a la pantalla principal
                Intent intent = new Intent(LoginActivity.this, MainActivity.class);
                startActivity(intent);
                finish();
            } else {
                Toast.makeText(LoginActivity.this, "Correo o contraseña incorrectos", Toast.LENGTH_SHORT).show();
            }
        });

        // Acción del botón de registrarse
        signupButton.setOnClickListener(v -> {
            // Redirigir al usuario a la actividad de registro
            Intent intent = new Intent(LoginActivity.this, SignUpActivity.class);
            startActivity(intent);
        });
    }

    // Método para autenticar al usuario (consultando la base de datos)
    private boolean authenticateUser(String email, String password) {
        boolean isAuthenticated = false;

        // Establecer conexión con la base de datos
        Connection connection = null;
        ResultSet resultSet = null;

        try {
            // Aquí conectas a tu base de datos usando JDBC o mediante una API REST
            // Ejemplo con JDBC:
            connection = DatabaseHelper.getConnection(); // Asegúrate de tener tu clase DatabaseHelper configurada

            // Realizar consulta SQL para verificar si el correo y la contraseña existen
            String query = "SELECT * FROM LOVELINK_USERS WHERE correo = ? AND contrasena = ?";
            PreparedStatement preparedStatement = connection.prepareStatement(query);
            preparedStatement.setString(1, email);
            preparedStatement.setString(2, password);
            resultSet = preparedStatement.executeQuery();

            // Si el resultado tiene un registro, significa que las credenciales son correctas
            if (resultSet.next()) {
                isAuthenticated = true;
            }

        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            // Cerrar la conexión a la base de datos
            try {
                if (resultSet != null) resultSet.close();
                if (connection != null) connection.close();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        return isAuthenticated;
    }
}

File: app/src/main/java/com/example/lovelink/LoginActivity.kt
package com.example.lovelink

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import com.example.lovelink.models.Cuenta
import com.example.lovelink.models.Usuario
import com.example.lovelink.network.CuentaApi
import com.example.lovelink.network.UsuarioApi
import com.example.lovelink.network.RetrofitClient
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.security.MessageDigest

class LoginActivity : Activity() {

    private lateinit var emailEditText: EditText
    private lateinit var passwordEditText: EditText
    private lateinit var loginButton: Button
    private lateinit var signupButton: Button

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        val prefs = getSharedPreferences("LoveLinkPrefs", MODE_PRIVATE)
        val isLoggedIn = prefs.getBoolean("isLoggedIn", false)
        val usuarioId = prefs.getLong("usuarioId", -1L)

        if (isLoggedIn && usuarioId != -1L) {
            val intent = Intent(this, PosiblesMatchesActivity::class.java)
            intent.putExtra("usuario_id", usuarioId)
            startActivity(intent)
            finish()
            return
        }

        emailEditText = findViewById(R.id.emailEditText)
        passwordEditText = findViewById(R.id.passwordEditText)
        loginButton = findViewById(R.id.loginButton)
        signupButton = findViewById(R.id.signupButton)

        loginButton.setOnClickListener {
            val email = emailEditText.text.toString().trim()
            val password = passwordEditText.text.toString().trim()

            if (email.isEmpty() || password.isEmpty()) {
                Toast.makeText(this, "Por favor, ingresa correo y contraseña", Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            }

            if (!android.util.Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
                Toast.makeText(this, "Correo no válido", Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            }

            if (password.length < 6) {
                Toast.makeText(this, "La contraseña debe tener al menos 6 caracteres", Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            }

            // 🔐 Aplicar hash SHA-256 antes de enviar
            val hashedPassword = hashSHA256(password)

            val cuenta = Cuenta(email = email, password = hashedPassword, telefono = "")

            val cuentaApi = RetrofitClient.retrofit.create(CuentaApi::class.java)
            cuentaApi.login(cuenta).enqueue(object : Callback<Cuenta> {
                override fun onResponse(call: Call<Cuenta>, response: Response<Cuenta>) {
                    if (response.isSuccessful && response.body() != null) {
                        val cuentaRespuesta = response.body()
                        val idCuenta = cuentaRespuesta?.id ?: -1L

                        if (idCuenta == -1L) {
                            Toast.makeText(this@LoginActivity, "Error al obtener ID de cuenta", Toast.LENGTH_SHORT).show()
                            return
                        }

                        getSharedPreferences("LoveLinkPrefs", MODE_PRIVATE).edit().apply {
                            putBoolean("isLoggedIn", true)
                            putString("email", email)
                            putLong("idCuenta", idCuenta)
                            apply()
                        }

                        val usuarioApi = RetrofitClient.retrofit.create(UsuarioApi::class.java)
                        usuarioApi.obtenerUsuarioPorCuenta(idCuenta).enqueue(object : Callback<Usuario> {
                            override fun onResponse(call: Call<Usuario>, response: Response<Usuario>) {
                                if (response.isSuccessful && response.body() != null) {
                                    val usuario = response.body()!!
                                    val usuarioId = usuario.id ?: -1L
                                    getSharedPreferences("LoveLinkPrefs", MODE_PRIVATE).edit().apply {
                                        putLong("usuarioId", usuarioId)
                                        apply()
                                    }
                                    val intent = Intent(this@LoginActivity, PosiblesMatchesActivity::class.java)
                                    intent.putExtra("usuario_id", usuarioId)
                                    startActivity(intent)
                                    finish()
                                } else {
                                    Toast.makeText(this@LoginActivity, "No se encontró el usuario", Toast.LENGTH_SHORT).show()
                                }
                            }

                            override fun onFailure(call: Call<Usuario>, t: Throwable) {
                                Toast.makeText(this@LoginActivity, "Error obteniendo usuario: ${t.message}", Toast.LENGTH_LONG).show()
                            }
                        })

                    } else {
                        Toast.makeText(this@LoginActivity, "Correo o contraseña incorrectos", Toast.LENGTH_SHORT).show()
                    }
                }

                override fun onFailure(call: Call<Cuenta>, t: Throwable) {
                    Toast.makeText(this@LoginActivity, "Error de conexión: ${t.localizedMessage}", Toast.LENGTH_LONG).show()
                }
            })
        }

        signupButton.setOnClickListener {
            val intent = Intent(this, SignUpActivity::class.java)
            startActivity(intent)
        }
        val forgotPasswordTextView = findViewById<TextView>(R.id.forgotPasswordTextView)
        forgotPasswordTextView.setOnClickListener {
            val intent = Intent(this, ForgotPasswordActivity::class.java)
            startActivity(intent)
        }

    }

    private fun hashSHA256(input: String): String {
        val bytes = MessageDigest.getInstance("SHA-256").digest(input.toByteArray())
        return bytes.joinToString("") { "%02x".format(it) }
    }
}
